File: lib/measurement.ts
import config from "../config/measurement.json" with { type: "json" };
import legacyConfig from "../config/scoring.json" with { type: "json" };

export const METHOD_VERSION = config.version;
export const MEASUREMENT = config;
export type Intent = keyof typeof config.intentLabels;
export type Level = keyof typeof config.levelLabels;
export type Classification = { intent: Intent; intentLabel: string; level: Level; levelLabel: string; chase: boolean; reportedChase: boolean; panic: boolean; bullish: boolean; bearish: boolean; evidence: string[]; reason: string };
export type Observation = { id: string; text: string; date: string; source: string; author: string; code: string; url: string; contentKind?: string };
export type ClassifiedPost = Observation & { classification: Classification };
export const round = (value: number) => Math.round(value * 10) / 10;
type Pattern = keyof typeof config.patterns;
const has = (key: Pattern, text: string) => new RegExp(config.patterns[key], "i").test(text);
const matches = (key: Pattern, text: string) => [...text.matchAll(new RegExp(config.patterns[key], "gi"))];

/** Context rules, not a trained model or a probability of a real trade. */
export function classifyText(input: string): Classification | null {
  const text = input.normalize("NFKC").replace(/\s+/g, " ").trim();
  if (text.length < 2 || text.length > 2000 || new RegExp(config.spamPatterns).test(text) || legacyConfig.spam.some(word => text.includes(word)) || legacyConfig.prose.filter(word => text.includes(word)).length >= 2) return null;
  const evidence: string[] = [];
  const intents = new Set<Intent>();
  let bullish = false, bearish = false, panic = false;
  const clauses = text.match(new RegExp(config.patterns.clauses, "g")) || [text];
  const ownContext = clauses.filter(clause => !has("thirdParty", clause) && !has("rhetorical", clause) && !has("historical", clause) && !has("refusal", clause)).join("，");
  for (const raw of clauses) {
    const clause = raw.trim();
    const related = ["chase", "buy", "rising", "urgent", "panic", "bullish", "bearish"].some(key => has(key as Pattern, clause));
    if (!related) continue;
    if (has("rhetorical", clause)) { intents.add("avoid"); evidence.push(`反问：${clause}`); continue; }
    if (has("historical", clause)) { intents.add("past"); evidence.push(`历史回顾：${clause}`); continue; }
    if (has("thirdParty", clause) && !has("following", clause)) { intents.add("quoted"); evidence.push(`他人观点：${clause}`); continue; }
    if (has("refusal", clause)) { intents.add("avoid"); evidence.push(`劝阻或观望：${clause}`); continue; }
    const positiveHits = (key: Pattern) => matches(key, clause).filter(match => !has("negation", clause.slice(Math.max(0, match.index! - 8), match.index)));
    const chase = positiveHits("chase").length > 0;
    const buy = positiveHits("buy").length > 0;
    const rising = has("rising", ownContext);
    const urgent = has("urgent", clause) && (buy || rising);
    // A wish for a limit-up is not an intention to buy at a higher price.
    if (chase || ((urgent || buy) && rising)) {
      if (has("question", clause) && !has("reported", clause)) intents.add("chase_question");
      else if (has("reported", clause) && !/如果|假如|打算|准备|明天|计划/.test(clause)) intents.add("reported_chase");
      else intents.add("chase_intent");
      evidence.push(`追涨语境：${clause}`);
    } else if (urgent && buy) { intents.add("aggressive_buy"); evidence.push(`激进买入，缺少追涨价格语境：${clause}`); }
    else if (buy) { intents.add("buy"); evidence.push(`买入表达：${clause}`); }
    else if (has("wish", clause) && positiveHits("rising").length) { intents.add("wish"); evidence.push(`看涨期待：${clause}`); }
    if (positiveHits("panic").length && !(has("plan", clause) && has("condition", text))) { panic = true; evidence.push(`恐慌表达：${clause}`); }
    if (positiveHits("bullish").length) bullish = true;
    if (positiveHits("bearish").length) bearish = true;
  }
  const order: Intent[] = ["reported_chase", "chase_intent", "chase_question", "aggressive_buy", "buy", "wish", "avoid", "past", "quoted"];
  const intent = order.find(item => intents.has(item)) ?? (panic ? "panic" : "discussion");
  let level: Level = "unknown";
  // The label is a style of the text, not the author's actual experience.
  const own = clauses.filter(clause => !has("thirdParty", clause) && !has("rhetorical", clause)).join("，");
  const hasOwn = (key: Pattern) => matches(key, own).some(match => !has("negation", own.slice(Math.max(0, match.index! - 8), match.index)));
  if (config.levelsEnabled) {
    if (hasOwn("basic")) level = "L1";
    else if (text.length >= 70 && hasOwn("research") && hasOwn("evidence") && hasOwn("comparison") && hasOwn("risk")) level = "L5";
    else if (text.length >= 35 && hasOwn("scenario") && matches("risk", own).length >= 2 && hasOwn("condition")) level = "L4";
    else if (text.length >= 20 && hasOwn("plan") && hasOwn("condition")) level = "L3";
    else if (has("following", text) && !has("refusal", text) && !has("rhetorical", text)) level = "L2";
  }
  if (level !== "unknown") evidence.push(`表达类型：${config.levelLabels[level]}（语境规则试验）`);
  return { intent, intentLabel: config.intentLabels[intent], level, levelLabel: config.levelLabels[level], chase: intent === "reported_chase" || intent === "chase_intent", reportedChase: intent === "reported_chase", panic, bullish, bearish, evidence: [...new Set(evidence)].slice(0, 6), reason: evidence.length ? "按语境区分行动、期待、转述和否定；自述不等于已核实成交。" : "缺少足够的行动或类型证据，保留为一般讨论／无法判断。" };
}

export function prepareObservations(rows: Observation[], date: string, cutoff = "23:59:59") {
  const seen = new Set<string>(), authorCounts = new Map<string, number>();
  const authors = new Set<string>();
  const analyzed: ClassifiedPost[] = [];
  let observedPosts = 0, unknownAuthorPosts = 0, excludedDate = 0, excludedNoise = 0;
  const ordered = [...rows].sort((a, b) => b.date.localeCompare(a.date) || b.source.localeCompare(a.source) || b.id.localeCompare(a.id));
  for (const row of ordered) {
    if (row.date.slice(0, 10) !== date || row.date.slice(11, 19) > cutoff) { excludedDate++; continue; }
    const classification = classifyText(row.text);
    if (!classification) { excludedNoise++; continue; }
    // Different accounts saying the same thing are still different observed
    // participants; do not erase a crowd by globally deduplicating its words.
    const identity = `${row.source}:${row.author || `missing:${row.id}`}:${row.text.toLowerCase().replace(/\s+/g, "")}`;
    if (seen.has(identity)) { excludedNoise++; continue; }
    seen.add(identity);
    observedPosts++;
    if (row.author) authors.add(`${row.source}:${row.author}`); else unknownAuthorPosts++;
    const authorKey = row.author ? `${row.source}:${row.author}` : `${row.source}:missing:${row.id}`;
    const count = authorCounts.get(authorKey) || 0;
    authorCounts.set(authorKey, count + 1);
    if (count < config.maxPostsPerAuthor) analyzed.push({ ...row, classification });
  }
  return { analyzed, observedPosts, observedAuthors: authors.size, unknownAuthorPosts, excludedDate, excludedNoise };
}

export function summarizeExpressions(posts: ClassifiedPost[]) {
  const total = posts.length;
  const rate = (count: number) => total ? round(count / total * 100) : null;
  const count = (key: "chase" | "reportedChase" | "panic" | "bullish" | "bearish") => posts.filter(post => post.classification[key]).length;
  const levels = (Object.keys(config.levelLabels) as Level[]).map(key => { const n = posts.filter(post => post.classification.level === key).length; return { key, label: config.levelLabels[key], count: n, share: rate(n) }; });
  const intents = (Object.keys(config.intentLabels) as Intent[]).map(key => { const n = posts.filter(post => post.classification.intent === key).length; return { key, label: config.intentLabels[key], count: n, share: rate(n) }; });
  return { sampleCount: total, chase: rate(count("chase")), reportedChase: rate(count("reportedChase")), panic: rate(count("panic")), bullish: rate(count("bullish")), bearish: rate(count("bearish")), direction: total ? round((count("bullish") - count("bearish")) / total * 100) : null, intensity: rate(posts.filter(post => post.classification.bullish || post.classification.bearish).length), levels, intents, levelCoverage: rate(total - levels.find(row => row.key === "unknown")!.count), l1l2Share: rate(levels.filter(row => row.key === "L1" || row.key === "L2").reduce((sum, row) => sum + row.count, 0)), thin: total < config.minimumSemanticSamples };
}
export type ExpressionSummary = ReturnType<typeof summarizeExpressions>;
export type BaselinePoint = { date: string; value: number | null; complete: boolean; universe: string; methodVersion: string; recordType?: string; cutoff: string };
export function historicalPercentile(value: number | null, history: BaselinePoint[], date: string, universe: string, complete: boolean, cutoff: string) {
  const byDate = new Map<string, BaselinePoint>();
  for (const row of history) if (row.date < date && row.complete && row.universe === universe && row.methodVersion === METHOD_VERSION && row.recordType === "measured" && row.cutoff === cutoff && row.value != null && Number.isFinite(row.value) && row.value >= 0) byDate.set(row.date, row);
  const past = [...byDate.values()].sort((a, b) => a.date.localeCompare(b.date)).slice(-config.baselineDays);
  const validValue = value != null && Number.isFinite(value) && value >= 0;
  const reason = !complete ? "当前采集覆盖不足" : !validValue ? "账户标识不足" : past.length < config.minimumBaselineDays ? "历史基线不足" : value === 0 ? "窗口内无有效讨论" : "历史分位";
  const ready = complete && validValue && past.length >= config.minimumBaselineDays;
  const score = ready ? value === 0 ? 0 : round(100 * past.reduce((sum, row) => sum + (row.value! < value! ? 1 : row.value === value ? 0.5 : 0), 0) / past.length) : null;
  const recent = past.slice(-20).map(row => row.value!).sort((a, b) => a - b);
  const median = ready ? (recent[Math.floor((recent.length - 1) / 2)] + recent[Math.floor(recent.length / 2)]) / 2 : null;
  const previous = complete && validValue ? past.at(-1) : undefined;
  return { score, baselineDays: past.length, minimumDays: config.minimumBaselineDays, reason,
    baselineMedian: median, growthPct: median != null && median > 0 ? round((value! / median - 1) * 100) : null,
    abnormalAttention: median != null ? Math.round(Math.log((1 + value!) / (1 + median)) * 1000) / 1000 : null,
    comparisonDate: previous?.date ?? null, previousAuthors: previous?.value ?? null,
    authorChange: previous ? value! - previous.value! : null, lowBase: median === 0 };
}

export function marketPrefix(code: string) { return /^(920|[48])/.test(code) ? "bj" : /^[569]/.test(code) ? "sh" : "sz"; }
export function universeKey(codes: string[], source = "eastmoney") { return `${config.scopeVersion}:${source}:${[...new Set(codes)].sort().join(",")}`; }

File: sentiment/measurement.py
"""V4 context rules and observable proportions, shared with TypeScript via JSON."""
from __future__ import annotations

import json
import math
import re
import unicodedata
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
CONFIG = json.loads((ROOT / "config/measurement.json").read_text(encoding="utf-8"))
LEGACY = json.loads((ROOT / "config/scoring.json").read_text(encoding="utf-8"))
METHOD_VERSION = CONFIG["version"]


def rounded(value):
    return math.floor(value * 10 + 0.5) / 10


def has(key, text):
    return re.search(CONFIG["patterns"][key], text, re.I) is not None


def matches(key, text):
    return list(re.finditer(CONFIG["patterns"][key], text, re.I))


def classify_text(text):
    text = re.sub(r"\s+", " ", unicodedata.normalize("NFKC", text)).strip()
    if len(text) < 2 or len(text) > 2000 or re.search(CONFIG["spamPatterns"], text) or any(word in text for word in LEGACY["spam"]) or sum(word in text for word in LEGACY["prose"]) >= 2:
        return None
    evidence, intents = [], set()
    bullish = bearish = panic = False
    clauses = re.findall(CONFIG["patterns"]["clauses"], text) or [text]
    own_context = "，".join(clause for clause in clauses if not any(has(key, clause) for key in ("thirdParty", "rhetorical", "historical", "refusal")))
    for raw in clauses:
        clause = raw.strip()
        if not any(has(key, clause) for key in ("chase", "buy", "rising", "urgent", "panic", "bullish", "bearish")):
            continue
        if has("rhetorical", clause):
            intents.add("avoid")
            evidence.append(f"反问：{clause}")
            continue
        if has("historical", clause):
            intents.add("past")
            evidence.append(f"历史回顾：{clause}")
            continue
        if has("thirdParty", clause) and not has("following", clause):
            intents.add("quoted")
            evidence.append(f"他人观点：{clause}")
            continue
        if has("refusal", clause):
            intents.add("avoid")
            evidence.append(f"劝阻或观望：{clause}")
            continue
        def positive(key):
            return [m for m in matches(key, clause) if not has("negation", clause[max(0, m.start() - 8):m.start()])]
        chase, buy = bool(positive("chase")), bool(positive("buy"))
        rising = has("rising", own_context)
        urgent = has("urgent", clause) and (buy or rising)
        if chase or ((urgent or buy) and rising):
            if has("question", clause) and not has("reported", clause):
                intents.add("chase_question")
            elif has("reported", clause) and not re.search("如果|假如|打算|准备|明天|计划", clause):
                intents.add("reported_chase")
            else:
                intents.add("chase_intent")
            evidence.append(f"追涨语境：{clause}")
        elif urgent and buy:
            intents.add("aggressive_buy")
            evidence.append(f"激进买入，缺少追涨价格语境：{clause}")
        elif buy:
            intents.add("buy")
            evidence.append(f"买入表达：{clause}")
        elif has("wish", clause) and positive("rising"):
            intents.add("wish")
            evidence.append(f"看涨期待：{clause}")
        if positive("panic") and not (has("plan", clause) and has("condition", text)):
            panic = True
            evidence.append(f"恐慌表达：{clause}")
        bullish = bullish or bool(positive("bullish"))
        bearish = bearish or bool(positive("bearish"))
    order = ["reported_chase", "chase_intent", "chase_question", "aggressive_buy", "buy", "wish", "avoid", "past", "quoted"]
    intent = next((key for key in order if key in intents), "panic" if panic else "discussion")
    own = "，".join(clause for clause in clauses if not has("thirdParty", clause) and not has("rhetorical", clause))
    def own_has(key):
        return any(not has("negation", own[max(0, m.start() - 8):m.start()]) for m in matches(key, own))
    level = "unknown"
    if CONFIG.get("levelsEnabled", False) and own_has("basic"):
        level = "L1"
    elif CONFIG.get("levelsEnabled", False) and len(text) >= 70 and all(own_has(key) for key in ("research", "evidence", "comparison", "risk")):
        level = "L5"
    elif CONFIG.get("levelsEnabled", False) and len(text) >= 35 and own_has("scenario") and len(matches("risk", own)) >= 2 and own_has("condition"):
        level = "L4"
    elif CONFIG.get("levelsEnabled", False) and len(text) >= 20 and own_has("plan") and own_has("condition"):
        level = "L3"
    elif CONFIG.get("levelsEnabled", False) and has("following", text) and not has("refusal", text) and not has("rhetorical", text):
        level = "L2"
    if level != "unknown":
        evidence.append(f"表达类型：{CONFIG['levelLabels'][level]}（语境规则试验）")
    return {"intent": intent, "intentLabel": CONFIG["intentLabels"][intent], "level": level,
            "levelLabel": CONFIG["levelLabels"][level], "chase": intent in ("reported_chase", "chase_intent"),
            "reportedChase": intent == "reported_chase", "panic": panic, "bullish": bullish, "bearish": bearish,
            "evidence": list(dict.fromkeys(evidence))[:6],
            "reason": "按语境区分行动、期待、转述和否定；自述不等于已核实成交。" if evidence else "缺少足够的行动或类型证据，保留为一般讨论／无法判断。"}


def prepare_observations(rows, day, cutoff="23:59:59"):
    seen, authors, counts, analyzed = set(), set(), {}, []
    observed = unknown = excluded_date = noise = 0
    for row in sorted(rows, key=lambda row: (row["date"], row["source"], row["id"]), reverse=True):
        if row["date"][:10] != day or row["date"][11:19] > cutoff:
            excluded_date += 1
            continue
        classification = classify_text(row["text"])
        if classification is None:
            noise += 1
            continue
        identity = row["source"] + ":" + (row["author"] or "missing:" + row["id"]) + ":" + re.sub(r"\s+", "", row["text"].lower())
        if identity in seen:
            noise += 1
            continue
        seen.add(identity)
        observed += 1
        if row["author"]:
            authors.add(row["source"] + ":" + row["author"])
        else:
            unknown += 1
        key = row["source"] + ":" + (row["author"] or "missing:" + row["id"])
        count = counts.get(key, 0)
        counts[key] = count + 1
        if count < CONFIG["maxPostsPerAuthor"]:
            analyzed.append({**row, "classification": classification})
    return {"analyzed": analyzed, "observedPosts": observed, "observedAuthors": len(authors), "unknownAuthorPosts": unknown, "excludedDate": excluded_date, "excludedNoise": noise}


def summarize_expressions(posts):
    total = len(posts)
    def rate(n):
        return rounded(n / total * 100) if total else None
    def count(key):
        return sum(bool(post["classification"][key]) for post in posts)
    levels = [{"key": key, "label": label, "count": sum(p["classification"]["level"] == key for p in posts)} for key, label in CONFIG["levelLabels"].items()]
    intents = [{"key": key, "label": label, "count": sum(p["classification"]["intent"] == key for p in posts)} for key, label in CONFIG["intentLabels"].items()]
    for row in levels + intents:
        row["share"] = rate(row["count"])
    return {"sampleCount": total, **{key: rate(count(key)) for key in ("chase", "reportedChase", "panic", "bullish", "bearish")},
            "direction": rounded((count("bullish") - count("bearish")) / total * 100) if total else None,
            "intensity": rate(sum(p["classification"]["bullish"] or p["classification"]["bearish"] for p in posts)),
            "levels": levels, "intents": intents, "levelCoverage": rate(total - levels[-1]["count"]),
            "l1l2Share": rate(sum(row["count"] for row in levels if row["key"] in ("L1", "L2"))),
            "thin": total < CONFIG["minimumSemanticSamples"]}


def historical_percentile(value, history, day, universe, complete, cutoff):
    by_date = {row["date"]: row for row in history if row["date"] < day and row.get("complete") and row.get("universe") == universe and row.get("methodVersion") == METHOD_VERSION and row.get("recordType") == "measured" and row.get("cutoff") == cutoff and isinstance(row.get("value"), (float, int)) and math.isfinite(row["value"]) and row["value"] >= 0}
    past = [by_date[key] for key in sorted(by_date)][-CONFIG["baselineDays"]:]
    valid_value = isinstance(value, (int, float)) and not isinstance(value, bool) and math.isfinite(value) and value >= 0
    reason = "当前采集覆盖不足" if not complete else "账户标识不足" if not valid_value else "历史基线不足" if len(past) < CONFIG["minimumBaselineDays"] else "窗口内无有效讨论" if value == 0 else "历史分位"
    ready = complete and valid_value and len(past) >= CONFIG["minimumBaselineDays"]
    score = (0 if value == 0 else rounded(100 * sum(1 if row["value"] < value else .5 if row["value"] == value else 0 for row in past) / len(past))) if ready else None
    recent = sorted(row["value"] for row in past[-20:])
    median = (recent[(len(recent)-1)//2] + recent[len(recent)//2]) / 2 if ready else None
    previous = past[-1] if complete and valid_value and past else None
    return {"score": score, "baselineDays": len(past), "minimumDays": CONFIG["minimumBaselineDays"], "reason": reason,
            "baselineMedian": median, "growthPct": rounded((value / median - 1) * 100) if median is not None and median > 0 else None,
            "abnormalAttention": math.floor(math.log((1 + value) / (1 + median)) * 1000 + .5) / 1000 if median is not None else None,
            "comparisonDate": previous["date"] if previous else None, "previousAuthors": previous["value"] if previous else None,
            "authorChange": value - previous["value"] if previous else None, "lowBase": median == 0}


def universe_key(codes, source="eastmoney"):
    return f"{CONFIG['scopeVersion']}:{source}:{','.join(sorted(set(codes)))}"
